app/src/test/java/com/example/android/architecture/blueprints/todoapp/LiveDataTestUtil.kt

```diff
@@ -23,26 +22,50 @@ import java.util.concurrent.CountDownLatch
 import java.util.concurrent.TimeUnit
 import java.util.concurrent.TimeoutException
 
+/**
+ * Gets the value of a [LiveData] or waits for it to have one, with a timeout.
+ *
+ * Use this extension from host-side (JVM) tests. It's recommended to use it alongside
+ * `InstantTaskExecutorRule` or a similar mechanism to execute tasks synchronously.
+ */
 @VisibleForTesting(otherwise = VisibleForTesting.NONE)
-fun <T> LiveData<T>.awaitNextValue(
+fun <T> LiveData<T>.getOrAwaitValue(
     time: Long = 2,
-    timeUnit: TimeUnit = TimeUnit.SECONDS
+    timeUnit: TimeUnit = TimeUnit.SECONDS,
+    afterObserve: () -> Unit = {}
 ): T {
     var data: T? = null
     val latch = CountDownLatch(1)
     val observer = object : Observer<T> {
         override fun onChanged(o: T?) {
             data = o
             latch.countDown()
-            this@awaitNextValue.removeObserver(this)
+            this@getOrAwaitValue.removeObserver(this)
         }
     }
     this.observeForever(observer)
+
+    afterObserve.invoke()
+
     // Don't wait indefinitely if the LiveData is not set.
     if (!latch.await(time, timeUnit)) {
+        this.removeObserver(observer)
         throw TimeoutException("LiveData value was never set.")
     }
 
     @Suppress("UNCHECKED_CAST")
     return data as T
```

Comment:
I think this will throw if you `ld.value = null`

Reply:
I can't find why would that happen. Can you expand?